flang/lib/Optimizer/Transforms/AddDebugInfo.cpp

```diff
@@ -45,13 +46,44 @@ namespace fir {
 namespace {
 
 class AddDebugInfoPass : public fir::impl::AddDebugInfoBase<AddDebugInfoPass> {
+  void handleDeclareOp(fir::cg::XDeclareOp declOp,
+                       mlir::LLVM::DIFileAttr fileAttr,
+                       mlir::LLVM::DIScopeAttr scopeAttr,
+                       fir::DebugTypeGenerator &typeGen, uint32_t &argNo);
+
 public:
   AddDebugInfoPass(fir::AddDebugInfoOptions options) : Base(options) {}
   void runOnOperation() override;
 };
 
+static uint32_t getLineFromLoc(mlir::Location loc) {
+  uint32_t line = 1;
+  if (auto fileLoc = mlir::dyn_cast<mlir::FileLineColLoc>(loc))
+    line = fileLoc.getLine();
+  return line;
+}
+
 } // namespace
 
+void AddDebugInfoPass::handleDeclareOp(fir::cg::XDeclareOp declOp,
+                                       mlir::LLVM::DIFileAttr fileAttr,
+                                       mlir::LLVM::DIScopeAttr scopeAttr,
+                                       fir::DebugTypeGenerator &typeGen,
+                                       uint32_t &argNo) {
+  mlir::MLIRContext *context = &getContext();
+  mlir::OpBuilder builder(context);
+
+  bool isLocal = (declOp.getMemref().getDefiningOp() != nullptr);
```

Comment:
I find `isLocal` a bit confusing here because global variables would also have "isLocal" set to true (since they are the result of a fir.address_of).

I guess reversing the logic and using "isDummyArgument" would described more what this is about. Looking for the existence of a defining op is also not solid because there are cases (e.g.) where there will be a little bit of argument "cooking" before the hlfir.declare (converts/unboxing). For instance in:

```
subroutine foo(c)
  character(10) :: c
end subroutine
```

Also, I am not sure what "argNo" ends-up being used for in the DWARF, but it is likely not accurate since hlfir.declare are created alphabetically (except when a variable "B" appears in the specification expression of another variable "A", in which case "B" declare is created before "A").
Does the position must be the Fortran argument position, or the one in the LLVM signature? These may differ since some Fortran function results may be passed by pointer via the first argument in the LLVM signature, and some Fortran arguments are split (The length of character arguments is passed after all Fortran arguments).

So we may need to add some info on the fir/hlfir declare to indicate something is a dummy argument, and what is its position.

Do you also know what should happen after inlining?

Reply:
The argNo value we pass will determine the order in which debugger will show the variables.

```
function fn1(c1, a1, b1) result (res1)
...
end function

(GDB) frame
#0  fn1 (c1=.FALSE., a1=1, b1=2.1000000000000001) at ../test1.f90:20

```
You were right that `DeclareOP` were in alphabetical order. I have now used `BlockArgument::getArgNumber` and it seems to return the correct order value. I am also using cast to BlockArgument to check if it is a dummy argument. But the problem of argument cooking in certain cases remain. I added a FIXME for that. Once I know all the cases where it can happen, it probably should not be too difficult to handle.

From inlining, you mean inlining that can happen before we process `DeclareOp` or after. If it happens after then I think it is up to that pass to adjust any metadata accordingly. Are there instances where inlining will happen before `AddDebugInfo` runs?

Reply:
MLIR inlining is not quite supported in FIR dialect, but you can experiment with some cases (some of them may not work) using `flang-new -O3 -mmlir -inline-all=true`, e.g. for this test:
```
subroutine test(x, y)
  real, target :: x, y
  x = y ! may alias
  call inner(x, y)
contains
  subroutine inner(x, y)
    real :: x, y
    x = y ! may not alias
  end subroutine inner
end subroutine test
```
I get this MLIR:
```
// -----// IR Dump After Inliner (inline) //----- //
module attributes {dlti.dl_spec = #dlti.dl_spec<#dlti.dl_entry<i1, dense<8> : vector<2xi64>>, #dlti.dl_entry<!llvm.ptr, dense<64> : vector<4xi64>>, #dlti.dl_entry<f80, dense<128> : vector<2xi64>>, #dlti.dl_entry<i32, dense<32> : vector<2xi64>>, #dlti.dl_entry<i16, dense<16> : vector<2xi64>>, #dlti.dl_entry<i8, dense<8> : vector<2xi64>>, #dlti.dl_entry<f128, dense<128> : vector<2xi64>>, #dlti.dl_entry<!llvm.ptr<271>, dense<32> : vector<4xi64>>, #dlti.dl_entry<f16, dense<16> : vector<2xi64>>, #dlti.dl_entry<!llvm.ptr<270>, dense<32> : vector<4xi64>>, #dlti.dl_entry<f64, dense<64> : vector<2xi64>>, #dlti.dl_entry<i128, dense<128> : vector<2xi64>>, #dlti.dl_entry<!llvm.ptr<272>, dense<64> : vector<4xi64>>, #dlti.dl_entry<i64, dense<64> : vector<2xi64>>, #dlti.dl_entry<"dlti.stack_alignment", 128 : i64>, #dlti.dl_entry<"dlti.endianness", "little">>, fir.defaultkind = "a1c4d8i4l4r4", fir.kindmap = "", fir.target_cpu = "x86-64", llvm.data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128", llvm.target_triple = "x86_64-unknown-linux-gnu"} {
  func.func @_QPtest(%arg0: !fir.ref<f32> {fir.bindc_name = "x", fir.target}, %arg1: !fir.ref<f32> {fir.bindc_name = "y", fir.target}) {
    %0 = fir.dummy_scope : !fir.dscope
    %1 = fir.declare %arg0 dummy_scope %0 {fortran_attrs = #fir.var_attrs<target>, uniq_name = "_QFtestEx"} : (!fir.ref<f32>, !fir.dscope) -> !fir.ref<f32>
    %2 = fir.declare %arg1 dummy_scope %0 {fortran_attrs = #fir.var_attrs<target>, uniq_name = "_QFtestEy"} : (!fir.ref<f32>, !fir.dscope) -> !fir.ref<f32>
    %3 = fir.load %2 : !fir.ref<f32>
    fir.store %3 to %1 : !fir.ref<f32>
    %4 = fir.dummy_scope : !fir.dscope
    %5 = fir.declare %1 dummy_scope %4 {uniq_name = "_QFtestFinnerEx"} : (!fir.ref<f32>, !fir.dscope) -> !fir.ref<f32>
    %6 = fir.declare %2 dummy_scope %4 {uniq_name = "_QFtestFinnerEy"} : (!fir.ref<f32>, !fir.dscope) -> !fir.ref<f32>
    %7 = fir.load %6 : !fir.ref<f32>
    fir.store %7 to %5 : !fir.ref<f32>
    return
  }
}
```

Reply:
Thanks. I looked at the IR in such case and it seems that function that gets inlined loses any `FuncOp`. Name in `DeclareOp` and `CallSiteLoc` may provide some clues but work will be required to retain enough information during inlining to enable `AddDebugInfo` to do its job. In any case, our first target should be to get the debug working with -O0.